src/lib/core/tt_sort.c

```diff
@@ -0,0 +1,515 @@
+		memcpy(sort->splitters + i * sort->elem_size,
+		       samples + si * sort->elem_size, sort->elem_size);
+	}
+
+	free(samples);
+}
+
+/**
+ * Check whether data part assigned to a thread is presorted.
+ */
+static int
+check_presorted(va_list ap)
+{
+	struct sort_worker *worker = va_arg(ap, typeof(worker));
+	struct tt_sort *sort = worker->sort;
+	worker->presorted = true;
+
+	void *pos = sort->data + worker->begin * sort->elem_size;
+	void *limit = sort->data + (worker->end - 1) * sort->elem_size;
+	for (; pos < limit; pos += sort->elem_size) {
+		if (sort->cmp(pos, pos + sort->elem_size, sort->cmp_arg) > 0) {
+			worker->presorted = false;
+			break;
+		}
+	}
+
+	return 0;
+}
+
+/**
+ * Sort all the data without dividing to buckets. Used in case sorting
+ * is limited to a single thread.
+ */
+static int
+sort_all(va_list ap)
+{
+	struct tt_sort *sort = va_arg(ap, typeof(sort));
+
+	qsort_arg_st(sort->data, sort->elem_count, sort->elem_size, sort->cmp,
+		     sort->cmp_arg);
+
+	return 0;
+}
+
+/**
+ * See usage description in header.
+ *
+ * The algorithm idea is to split the data into buckets, each element in
+ * bucket `i` is greater than each element in bucket `i - 1`, and then
+ * sort the buckets. As a result we will sort original array.
+ *
+ * So the algo outline is next:
+ *
+ * 1. Find buckets boundaries (splitters).
+ * 2. For each element in data find to what bucket is belongs to.
+ * 3. Copy elements to their buckets.
+ * 4. Sort buckets using sequentional qsort_arg and then copy buckets back
+ *    to the original array.
+ *
+ * Steps 2, 3 and 4 are run in parallel on `thread_count` threads. Step 1
+ * does not have high computation cost and is run single thread.
+ *
+ * Additionally we check if data is already sorted before applying main algo.
+ *
+ * See also:
+ * [1] https://en.wikipedia.org/wiki/Samplesort
+ * [2] Super Scalar Sample Sort, Peter Sanders and Sebastian Winkel
+ *
+ * Although this implementation does not use superscalar approach to map
+ * elements to buckets as in [2] and use usual binary search.
+ */
+void
+tt_sort(void *data, size_t elem_count, size_t elem_size,
+	tt_sort_compare_f cmp, void *cmp_arg, int thread_count)
+{
+	struct tt_sort sort;
+	double b, e;
+
+	say_verbose("start sort, data size: %zu, elem size: %zu, threads: %d",
+		    elem_count, elem_size, thread_count);
+	if (elem_count < NOSPAWN_SIZE_THESHOLD) {
+		say_verbose("data size is less than threshold %d,"
+			    " sort in caller thread", NOSPAWN_SIZE_THESHOLD);
+		qsort_arg_st(data, elem_count, elem_size, cmp, cmp_arg);
+		return;
+	}
+
+	/*
+	 * Upper limit because element to bucket map has unsigned char
+	 * storage per element.
+	 */
+	assert(thread_count > 0 && thread_count <= TT_SORT_THREADS_MAX);
+
+	sort.data = data;
+	sort.elem_count = elem_count;
+	sort.elem_size = elem_size;
+	sort.cmp = cmp;
+	sort.cmp_arg = cmp_arg;
+	sort.thread_count = thread_count;
+
+	if (thread_count == 1) {
```

Comment:
This function (`tt_sort`) is too big. It seems that if we create a couple of helpers (for example `tt_sort_one_thread`) it would be much shorter and thus understandable.